src/mongo/s/stats/joom_latency_histogram.cpp

```diff
@@ -0,0 +1,123 @@
+#include "mongo/platform/basic.h"
+
+#include "mongo/s/stats/joom_latency_histogram.h"
+
+#include "mongo/bson/bsonobjbuilder.h"
+#include "mongo/platform/bits.h"
+
+namespace mongo {
+
+const std::array<uint64_t, JoomLatencyHistogram::kMaxBuckets>
+    JoomLatencyHistogram::kUpperBoundsMicros = {
+        1000,
+        5000,
+        10000,
+        20000,
+        50000,
+        100000,
+        300000,
+        500000,
+        1000000,
+        3000000,
+    };
+
+void JoomLatencyHistogram::increment(CommandName cmdName, uint64_t latency, bool isError) {
+    switch (cmdName) {
+        case CommandName::Insert:
+            _incrementData(latency, isError, &_insert);
+            break;
+        case CommandName::Find:
+            _incrementData(latency, isError, &_find);
+            break;
+        case CommandName::FindAndModify:
+            _incrementData(latency, isError, &_findAndModify);
+            break;
+        case CommandName::Update:
+            _incrementData(latency, isError, &_update);
+            break;
+        case CommandName::Delete:
+            _incrementData(latency, isError, &_delete);
+            break;
+        case CommandName::Aggregate:
+            _incrementData(latency, isError, &_aggregate);
+            break;
+        case CommandName::Distinct:
+            _incrementData(latency, isError, &_distinct);
+            break;
+        case CommandName::Other:
+            _incrementData(latency, isError, &_other);
+    }
+}
+
+void JoomLatencyHistogram::append(BSONObjBuilder* builder) const {
+    _append(&_insert, "insert", builder);
+    _append(&_find, "find", builder);
+    _append(&_findAndModify, "findAndModify", builder);
+    _append(&_update, "update", builder);
+    _append(&_delete, "delete", builder);
+    _append(&_aggregate, "aggregate", builder);
+    _append(&_distinct, "distinct", builder);
+    _append(&_other, "other", builder);
+}
+
+// Binary search for a bucket corresponding to a value. Each bucket is a upper bound and 
+// values must be less or equal to fit into. Values bigger than kUpperBoundsMicros[kMaxBuckets-1]
+// don't fit into any bucket and count separately. 
+int JoomLatencyHistogram::_getBucket(uint64_t value) {
```

Comment:
We can use `std::lower_bound` here. But I also find it amusing to stretch the muscles and implement something as classical as binary search :)

Reply:
Oh, well... I totally forgot about these functions:) Replaced with `std::lower_bound`. Now `upper bound` term in comment is a bit misleading:)